### ISSUE TITLE
Directed Acyclic Graphs (DAG)

### ISSUE THREAD

**Issue by techniq** [techniq/layerchart#30]
## Data model / API
- https://graphology.github.io/
- https://github.com/dagrejs/graphlib

## Layout
- [dagre](https://github.com/dagrejs/dagre)
  - https://github.com/dagrejs/dagre/wiki
  - https://github.com/dagrejs/dagre-d3
  - https://observablehq.com/@mbostock/hello-dagre
  - https://observablehq.com/@mbirk/dagre
  - https://observablehq.com/@valnik/dagre-3d
  - [examples](https://github.com/dagrejs/dagre-d3/wiki)
    - https://dagrejs.github.io/project/dagre-d3/latest/demo/clusters.html
- [elkjs](https://github.com/kieler/elkjs)
  - https://rtsys.informatik.uni-kiel.de/elklive/index.html
- [webcola](https://ialab.it.monash.edu/webcola)
- https://github.com/erikbrinkman/d3-dag
  - https://observablehq.com/@erikbrinkman/d3-dag-topological
  - https://observablehq.com/@erikbrinkman/d3-dag-sugiyama
  - [d3-dag with a big CTE (Svelte REPL)](https://svelte.dev/repl/83b44d5d50be4b8bb65d55cea1a5d1f1?version=3.49.0)

## Rendering
- https://www.sigmajs.org/
  - uses graphology for data model
- [cytoscape](https://js.cytoscape.org/)
  - supports many different [layout extensions](https://js.cytoscape.org/#extensions/layout-extensions) including
    - cola
    - d3-force
    - dagre
- https://memgraph.com/blog/you-want-a-fast-easy-to-use-and-popular-graph-visualization-tool
- https://reagraph.dev/?path=/docs/docs-getting-started-layouts--docs
- https://graphviz.org/

## Examples
- https://datavizcatalogue.com/methods/flow_chart.html
- https://en.wikipedia.org/wiki/Business_Process_Model_and_Notation
- https://js.cytoscape.org/#demos
- https://svelteflow.dev/examples/layout/dagre
  - also supports Elkjs
- https://graphviz.org/

**Comment by techniq** [comment]
Also, based on this notice...

![image](https://github.com/techniq/layerchart/assets/177476/32c87336-3679-44bb-a656-92fd812611a2)

but be worth looking at [graphology](https://www.npmjs.com/package/graphology) or [sigma](https://www.npmjs.com/package/sigma), although d3-dag might be sufficient for now

**Comment by regexident** [comment]
There is also [webcola](https://ialab.it.monash.edu/webcola) (aka cola.js), which provides support for an [alternative layout algorithm](https://ialab.it.monash.edu/webcola/examples/downwardedges.html) based on (afaik)

> Dwyer, Tim & Koren, Yehuda. (2005). DIG-COLA: Directed Graph Layout through Constrained Energy Minimization.. Proceedings of IEEE Symposium on Information Visualization. 9. 10.1109/INFVIS.2005.1532130. 

… and [much more](https://ialab.it.monash.edu/webcola/examples/).

**Comment by techniq** [comment]
Thanks @regexident, I forgot about `webcola`.

With that said, I do like the visual output of the different d3-dag layouts, but it could just come down to examples created.  I prefer to have a stable layout vs a simulation most of the time, although a simulation can be made to look "static" most of the time.

It's always tricky to find the best solution that is both easy to use for the 80-90% use cases, but can hopefully approach handling 100% of the use cases if needed).  We can also provide multiple layout components that use a different library/algorithm/layout/approach if needed.

I haven't really dived deep into this space to understand the pro/cons for each (simplicity, library size, performance, etc).

**Comment by regexident** [comment]
One difficulty I would with d3-dag is its need to provide node sizes upfront, which can be tricky if your nodes are not of a predetermined size, but e.g. depend on whatever your styles for them define.

**Comment by techniq** [comment]
Yeah, that is typically a challenge with the more "static" layouts (same with the d3-hierarchy layouts).

**Comment by mcmxcdev** [comment]
Just putting it out there that there is https://svelteflow.dev too, which focuses 100% on creating a DAG.

**Comment by techniq** [comment]
@mcmxcdev Svelte Flow is awesome for sure, and would be the better choice depending on the needs/constraints (editor vs data visualization).

Ironically (or coincidentally), I had the recent need for a DAG layout, and a few days ago started working on a new [dagrejs](https://github.com/dagrejs/dagre)-based layout component (similar to `<Tree>`, `<Sankey>`, etc.

Here's a quick video preview:

https://github.com/user-attachments/assets/c05c6d81-7e53-493c-a7ce-4ff3eeda4e08

Also adding a new [<Marker>](https://github.com/techniq/layerchart/issues/8) component for adding arrows to start/end (and mid) points, which will be nice for charts beyond a DAG graph (line charts, etc)

**Comment by techniq** [comment]
[Dagre](https://www.layerchart.com/docs/examples/Dagre) has been added.  Here is a more complex example:

https://github.com/user-attachments/assets/b99d3b13-27f0-4a19-8a44-3902c907d74d

You can also leverage [ForceSimulation](https://www.layerchart.com/docs/components/ForceSimulation) or [Sankey](https://www.layerchart.com/docs/components/Sankey) (although ForceSimulation currently has a [performance regression](https://github.com/techniq/layerchart/issues/451) in later versions of Svelte 5 that needs addressed)
